### ISSUE TITLE
Contains query does not work with composite key

### ISSUE THREAD

**Issue by sebastianfilke** [aspnet/EntityFrameworkCore#10384]
I have an Entity called QuestionnaireEntity and a composite key 

`builder.HasKey(e => new {e.IdQuestionnaire, e.IdQuestionnaireVersion});`

I want to do a select statement with in clause like this:

```
var idQuestionnaires = questionnaireModels.Select(x => new {x.IdQuestionnaire, x.IdQuestionnaireVersion});
var questionnaireEntities = Set<QuestionnaireEntity>().Where(x => idQuestionnaires.Contains(new {x.IdQuestionnaire, x.IdQuestionnaireVersion}))
                .ToList();
```
which results in the following exception:
```
2017/11/23 10:22:13.566 | Error | An exception occurred in the database while iterating the results of a query for context type 'Qkiss.Data.Context.PersistenceContext'.
System.InvalidOperationException: No mapping to a relational type can be found for the CLR type 'Expression[]'.
   at Microsoft.EntityFrameworkCore.Storage.RelationalTypeMapperExtensions.GetMapping(IRelationalTypeMapper typeMapper, Type clrType)
   at Microsoft.EntityFrameworkCore.Storage.RelationalTypeMapperExtensions.GetMappingForValue(IRelationalTypeMapper typeMapper, Object value)
   at Microsoft.EntityFrameworkCore.Query.Sql.DefaultQuerySqlGenerator.GetTypeMapping(Object value)
   at Microsoft.EntityFrameworkCore.Query.Sql.DefaultQuerySqlGenerator.VisitConstant(ConstantExpression constantExpression)
   at System.Linq.Expressions.ConstantExpression.Accept(ExpressionVisitor visitor)
   at Remotion.Linq.Parsing.ThrowingExpressionVisitor.Visit(Expression expression)
   at Microsoft.EntityFrameworkCore.Query.Sql.DefaultQuerySqlGenerator.VisitIn(InExpression inExpression)
   at Microsoft.EntityFrameworkCore.Query.Expressions.InExpression.Accept(ExpressionVisitor visitor)
   at Remotion.Linq.Parsing.ThrowingExpressionVisitor.Visit(Expression expression)
   at Microsoft.EntityFrameworkCore.Query.Sql.DefaultQuerySqlGenerator.GeneratePredicate(Expression predicate)
   at Microsoft.EntityFrameworkCore.Query.Sql.DefaultQuerySqlGenerator.VisitSelect(SelectExpression selectExpression)
   at Microsoft.EntityFrameworkCore.Query.Expressions.SelectExpression.Accept(ExpressionVisitor visitor)
   at Remotion.Linq.Parsing.ThrowingExpressionVisitor.Visit(Expression expression)
   at Microsoft.EntityFrameworkCore.Query.Sql.DefaultQuerySqlGenerator.GenerateSql(IReadOnlyDictionary`2 parameterValues)
   at Microsoft.EntityFrameworkCore.Query.Internal.ShaperCommandContext.GetRelationalCommand(IReadOnlyDictionary`2 parameters)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryingEnumerable`1.Enumerator.BufferlessMoveNext(Boolean buffer)
   at Microsoft.EntityFrameworkCore.ExecutionStrategyExtensions.<>c__DisplayClass12_0`2.<Execute>b__0(DbContext c, TState s)
   at Microsoft.EntityFrameworkCore.Storage.Internal.NoopExecutionStrategy.Execute[TState,TResult](TState state, Func`3 operation, Func`3 verifySucceeded)
   at Microsoft.EntityFrameworkCore.ExecutionStrategyExtensions.Execute[TState,TResult](IExecutionStrategy strategy, Func`2 operation, Func`2 verifySucceeded, TState state)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryingEnumerable`1.Enumerator.MoveNext()
   at Microsoft.EntityFrameworkCore.Query.Internal.LinqOperatorProvider.<_TrackEntities>d__17`2.MoveNext()
   at Microsoft.EntityFrameworkCore.Query.Internal.LinqOperatorProvider.ExceptionInterceptor`1.EnumeratorExceptionInterceptor.MoveNext()
Exception thrown: 'System.InvalidOperationException' in Microsoft.EntityFrameworkCore.dll
Exception thrown: 'System.InvalidOperationException' in System.Private.CoreLib.dll
Exception thrown: 'System.InvalidOperationException' in System.Private.CoreLib.dll
An exception of type 'System.InvalidOperationException' occurred in System.Private.CoreLib.dll but was not handled in user code
No mapping to a relational type can be found for the CLR type 'Expression[]'.
```

The following works perfect with single attributes and not composite keys:
```
var idCustomers = customerModels.Select(x => x.IdCustomer);
var customerEntities = Set<CustomerEntity>().Where(x => idCustomers.Contains(x.IdCustomer)).ToList();
```

So my question is: How can i obtain a query with select in clause and multiple attributes respectively composite keys?

**Comment by sebastianfilke** [comment]
I also tried the following. which seems to work but produces a warning:

```
var idQuestionnaires = questionnaireModels.Select(x => new Tuple<int, int>(x.IdQuestionnaire, x.IdQuestionnaireVersion));
var questionnaireEntities = Set<QuestionnaireEntity>()
                .Where(x => idQuestionnaires.Any(q => q.Item1.Equals(x.IdQuestionnaire) && q.Item2.Equals(x.IdQuestionnaireVersion))).ToList();
```
```
2017/11/23 10:36:18.056 | Warn | The LINQ expression 'where {from Tuple`2 q in __idQuestionnaires_0 where ([q].Item1.Equals([x].IdQuestionnaire) AndAlso [q].Item2.Equals([x].IdQuestionnaireVersion)) select [q] => Any()}' could not be translated and will be evaluated locally.
2017/11/23 10:36:18.056 | Warn | The LINQ expression 'where ([q].Item1.Equals([x].IdQuestionnaire) AndAlso [q].Item2.Equals([x].IdQuestionnaireVersion))' could not be translated and will be evaluated locally.
2017/11/23 10:36:18.083 | Warn | The LINQ expression 'Any()' could not be translated and will be evaluated locally.
```

**Comment by smitpatel** [comment]
Possible duplicate of #9424 & #10081 
What is the version of EF Core you are using?
Warnings will be there because we have to client eval it for now.

**Comment by sebastianfilke** [comment]
Yeah, it's somehow related to it but i think it's not exactly the same. i am using the newest version 2.0.1.
Also i am not using Tuple against Tuple. I am taking the items out of the Tuple and compare them. Which already works.

But the main problem for me is that i would expect that this works:

```
var idQuestionnaires = questionnaireModels.Select(x => new {x.IdQuestionnaire, x.IdQuestionnaireVersion});
var questionnaireEntities = Set<QuestionnaireEntity>().Where(x => idQuestionnaires.Contains(new {x.IdQuestionnaire, x.IdQuestionnaireVersion}))
                .ToList();
```

If i configure a composite key in my IEntityTypeConfiguration like this `builder.HasKey(e => new {e.IdQuestionnaire, e.IdQuestionnaireVersion});` i would expect that i can query i like this and not with a Tuple. So if it's the same for you close it. Otherwise i would say it's a little bit different.

**Comment by sebastianfilke** [comment]
Little addition: why i would except that this works is that a regular linq query does work:

```
var questionnaireEntity = questionnaireEntities.FirstOrDefault(x =>
                    new {x.IdQuestionnaire, x.IdQuestionnaireVersion}.Equals(
                        new {questionnaireModel.IdQuestionnaire, questionnaireModel.IdQuestionnaireVersion}));
```

**Comment by Tarig0** [comment]
Would it work if you did
```c#
///var idQuestionnaires = questionnaireModels.Select(x => new {x.IdQuestionnaire, x.IdQuestionnaireVersion});
var questionnaireEntities = Set<QuestionnaireEntity>().Where(x => idQuestionnaires.Contains(idQuestionnaires ))
                .ToList();
```

**Comment by ajcvickers** [comment]
Notes from triage: we will investigate for 2.1 fixing the exception and performing client eval instead. After that, probably translating this to SQL will go on the backlog.

**Comment by smitpatel** [comment]
Fixed in #9522
